nrf24L01_code/bootloader/WriteApp.c

```diff
@@ -0,0 +1,383 @@
+/*
+ * WriteApp.c
+ *
+ * Created: 11/28/2019 9:38:05 PM
+ *  Author: Adi
+ */ 
+#include "WriteApp.h"
+#include "radio_fptr.h"
+#include "GenericLib.h"
+
+
+#define BUFFER_LENGTH				((uint8_t) 32)
+#define PAGE_SIZE					((uint8_t) SPM_PAGESIZE)
+#define MSG_REC_TIMEOUT				((uint32_t) 100000000)
+#define MSG_RECEIVED				((uint8_t) 0xAA)
+#define MSG_NO_DATA					((uint8_t) 0x00)
+
+
+
+
+
+// Flashing state machine
+#define BOOTLOADER_INIT_NRF			((uint8_t) 'A')
+#define BOOTLOADER_FLASH_INIT		((uint8_t) 'S')
+#define BOOTLOADER_REC_DATA			((uint8_t) 'D')
+#define BOOTLOADER_FLASH_DATA		((uint8_t) 'F')
+#define BOOTLOADER_FLASH_PAGE_DONE	((uint8_t) 'G')
+#define BOOTLOADER_CHECK_CKS		((uint8_t) 'H')
+
+#define BOOTLOADER_FLASH_END		((uint8_t) 'K')
+#define BOOTLOADER_FLASH_ERROR		((uint8_t) 'L')
+
+
+// Flashing commands
+#define COMM_GET_STATE				((uint8_t) 'q')
+#define COMM_START_FLASH_PROCEDURE	((uint8_t) 'w')
+#define COMM_SEND_CHUNCK			((uint8_t) 'e')
+#define COMM_FLASH_PAGE				((uint8_t) 'r')
+#define COMM_NEXT_PAGE				((uint8_t) 't')
+#define COMM_CHECK_CKS				((uint8_t) 'y')
+#define COMM_FINISH_FLASH			((uint8_t) 'u')
+
+
+
+#define COMM_ABORT					((uint8_t) 'o')
+#define COMM_NO_COMMAND				((uint8_t) 'p')
+
+char bootloader_state = BOOTLOADER_INIT_NRF;
+
+
+typedef struct rxDataStruct
+{
+	uint8_t data[BUFFER_LENGTH];
+	uint8_t command;
+	uint8_t length;
+	uint8_t hasMessage;
+}rxDataStruct;
+
+rxDataStruct rxData;
+uint8_t flashPipe = 0;
+
+
+void rx_handler(uint8_t pipe, uint8_t * data, uint8_t payload_length) {
+	rxData.command = data[0];
+	memcpy(rxData.data, data + 1, payload_length - 1);	
+	rxData.length = payload_length - 1;	
+	rxData.hasMessage = MSG_RECEIVED;
+}
+void tx_handler(radio_tx_status tx_status) {
+	
+}
+
+
+
+void initNrf(uint8_t * rx_address)
+{
+	radio_config cfg = {
+		RADIO_ADDRESS_5BYTES,
+		RADIO_RETRANSMIT_WAIT_3000US,
+		RADIO_RETRANSMIT_15,
+		CHANNEL_112,
+		RADIO_250KBIT,
+		RADIO_CRC2_ENABLED,
+		RADIO_COUNT_WAVE_DISABLED,
+		RADIO_HIGHEST_0DBM,
+		RADIO_DYNAMIC_PAYLOAD_ENABLED,
+		RADIO_ACK_PAYLOAD_ENABLED,
+		RADIO_DYNAMIC_ACK_DISABLED,
+		RADIO_BOOTLOADER
+	};
+	__nrfRadio_Init(cfg);
+	
+	pipe_config pipe_cfg0 =
+	{
+		RADIO_PIPE0,
+		rx_address,
+		5,
+		RADIO_PIPE_RX_ENABLED,
+		RADIO_PIPE_AA_ENABLED,
+		RADIO_PIPE_DYNAMIC_PYALOAD_ENABLED
+	};
+	__nrfRadio_PipeConfig(pipe_cfg0);
+	
+	__nrfRadio_SetRxCallback(rx_handler);
+	__nrfRadio_SetTxCallback(tx_handler);
```

Comment:
if we don't use asynchronously transmission, then we don't need a TX callback - the NRF library should handle that NULL pointer (at least I'm hoping it does)